README.md
## JSON AND HTML GENERATOR

This is a simple CMS application to generate json and html. This Json is used to generate static html files in another project, exactly including this one [photoBlog](https://github.com/tomik23/photoBlog)

> The problem that was solved, I had a script that generated an empty `Json` with only added photos, and the rest I had to enter manually. It was not easy because everything had to be inserted and formatted manually.

It is now possible to save html files after clicking the `"NEW"` and `"UPDATE"` buttons. Of course, html are very primitive, but you can build on your own just modify ```sources > views> html> * .pug```

A `.env` file has been added in root with which we can control the port at which the page will be displayed to us and whether the checkbox (HTML SAVE) will be selected by default in the cms preview - `HTML_GENERATOR=true`

## Initialization

Before the first use, clone this repository and install node dependencies: `yarn` or `npm install`

## Basic configuration

The application needs folders `images`, 
This folder should contain folders with individual photos. See the `images` folder in this repository. 

> Note the folder name is the name of the `html` and `json` file that will be generated. The name should not contain special characters or spaces. A valid example of `one-day-in-berlin`.

## Run the app

Run the app, just call: `yarn start` or `npm run start`

Open the address `http://localhost:3000`. A list of available folders should appear on this page.

## TODO

- [x] first adding a page with a list of available folders with photos to choose from
- [x] the possibility to update the generated json
- [x] adding grids to better sort
- [x] the ability to move photos (drag-and-drop), the problem occurs when the images are from different cameras, or the file names are mixed up
- [x] html generator
- [x] small wysiwyg html editor

## Other information

Most important part of it's technology stack:

* [express.js](https://expressjs.com/)
* [pug](https://github.com/pugjs/pug)
* [flatpickr](https://github.com/flatpickr/flatpickr)
* [zooom.js](https://github.com/tomik23/zooom.js)
* [pell](https://github.com/jaredreich/pell)
* [SortableJS](https://github.com/SortableJS/Sortable)

---
### MAIN PAGE

<img src="screenshot/page.png">

---
### OPEN THE GENERATED HTML FILE

<img src="screenshot/page-html.png">

---
### CALENDAR

<img src="screenshot/calendar.png">

---
### WYSIWYG TEXT EDITOR

<img src="screenshot/wysiwyg.png">

---
> Each of the photos, including alt and photo description, can be moved among themselves to help manage the order of photos. Of course, the easiest way to do this is when `GRID` mode is enabled. 
### GRID-OFF

<img src="screenshot/grid-off.png">

---
### GRID-ON

<img src="screenshot/grid-on.png">

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## 2020-06-14
## Changed
- Removal of the semantic-ui library


## 2020-01-01
## Changed
- Placing Semantic-UI styles and fonts locally

## 2019-12-30
## Changed
- Update devDependencies
- Fixed missing br in captions under the pictures in html 

## 2019-11-28
### Changed
- Style repair with enlarged contenteditable field
- Support for the new line in the photos description
- Fixed issue width HTML_GENERATOR

## 2019-11-19
### Changed
- Separator in contenteditable div
- Change authorarticle to author so that the variable from .env can be read correctly

## 2019-11-18
### Added
- Adding the author field, default setting in the `.env AUTHOR=your name` configuration file

### Changed
- The publication date is no longer overwritten with the update date
- Change the line-height of the main text field when this field is enlarged to fill the entire screen

## 2019-11-12
### Change
- Photo path repair when creating new json - "NEW"

### Added
- WYSIWYG text editor
- Expanding WYSIWYG capabilities by clearing formatting and enlarging the main text field.

## 2019-11-08
### Changed
- Save html files, dependent on .env (HTML_GENERATOR=true/false) and checkbox (HTML SAVE)

## 2019-10-18
### Added
- Installation dotenv
- Create configuration files .env, config.js
- Function checking and creating main html/data folders
- Ability to generate html files

### Changed 
- Update `screenshot`
- Update README.md
- Remove html branche

## 2019-10-02
### Changed
- Delete unnecessary folder in photos
- Update devDependencies

## 2019-08-10
### Added
- PUG template

### Changed
- server.js -> app.js
- Return to vivid colors

### Removed
- HTML template
- express-es6-template-engine library

## 2019-08-07
### Added
- New sorting library - sortablejs and zoomjs. The possibility of enlarging and changing the position of photos.
- Displaying photos in the grid for better UX when sorting
- Button smooth-scroll to top of page
- Date of last modification of the json file under the title

### Changed
- Adding 'min' to zooom.js and zooom.css

### Removed
- Draggable.js library incompatible with zoom.js

## 2019-08-05
### Added 
- Responsive section "LIST OF PHOTOS"
- Caching vendor - css, js
- Drag-and-drop in the photos section

### Removed
- Conflict between zoom.js and drag-and-drop, zoom has been removed

## 2019-08-03
### Added
- Possibility to modify the already generated json file
- Support for 400 and 500 server status
- Picture caching - 1h

### Changed
- Change the appearance of the main page

## 2019-08-01
### Added
- First adding a page with a list of available folders with photos to choose from
- Preparation of the form with the possibility of adding seo elements as well as page content
- Calendar to the field with the date
- Zooom.js library with the possibility of enlarging photos
- Favicon support
- Semantic.min.css styles
- Form construction

server.js
const { existsSync, statSync } = require('fs');
const path = require('path');
const express = require('express');
const app = express();
const compression = require('compression');
const favicon = require('serve-favicon');

const { port, htmlGenerator, author } = require('./config');

const {
  getAllFiles,
  getAllDirectory,
  getAllJson,
  readJson,
  checkFolders,
} = require('./sources/helper/images');

const generate = htmlGenerator === 'true';

// compression static files
app.use(compression());

const fileTemplate = require('./sources/helper/template');
const fileHtml = require('./sources/helper/html');

// time for caching - maxAge
const oneHour = 60 * 1000 * 60;

const locals = {
  title: 'error',
  header: 'UPS !!! error occurred. The reason may be the wrong address.',
  info: 'The address should contain the name of the folder with photos.',
  address: `http://localhost:${port}/<b>name/folder-name</b>`,
};

// load View Engine
app.set('views', path.join(__dirname, 'sources/views'));
app.set('view engine', 'pug');

// form
app.use(express.urlencoded({ extended: false }));

// path to css, images, favico and html
app.use('/vendor', express.static(path.join(__dirname, 'sources/vendor')));
app.use('/html', express.static(path.join(__dirname, 'html')));
app.use('/images', express.static(path.join(__dirname, 'images'), { maxAge: oneHour }));

// json data
app.use('/update', express.static(path.join(__dirname, 'data')));

// favicon
app.use(favicon(path.join(__dirname, '/public/images/favicon.ico')));

// get post
app.post('/', (req, res) => {
  const {
    folderName,
    seoTitle,
    seoDescription,
    bodyTitle,
    bodyDate,
    bodyText,
    bodyAuthor,
    imageName,
    imageAlt,
    imageText,
    html,
  } = req.body;

  const imagePath = imageName.map((image, index) => `
    {
      "path": "images/${folderName}/",
      "img": "${image}",
      "alt": "${imageAlt[index]}",
      "text": ${JSON.stringify(imageText[index].replace(/\n?\r\n/g, '<br>'))} 
    }`);


  const config = {
    nameFolder: folderName,
    seoTitle: JSON.stringify(seoTitle),
    seoDescription: JSON.stringify(seoDescription),
    bodyTitle: JSON.stringify(bodyTitle),
    bodyDate: JSON.stringify(bodyDate),
    bodyText: JSON.stringify(bodyText),
    bodyAuthor: JSON.stringify(bodyAuthor),
    images: imagePath,
  };

  // save html files, dependent on .env (HTML_GENERATOR=true/false) and checkbox (HTML SAVE)
  if (generate || html === 'on') {
    // save html files
    fileHtml(config);
  }

  // save json filess
  fileTemplate(config);

  res.redirect('./success');
});

// checking if folders exist if they are not created
checkFolders('./data/');
checkFolders('./html/');

const keys = getAllDirectory('./images/');
const jsonFiles = getAllJson('./data/', 'json');
const htmlFiles = getAllJson('./html/', 'html');

const merged = keys.reduce((obj, key) => ({
  ...obj,
  [key]: jsonFiles.includes(key),
}), {});

const allFolders = Object.entries(merged).map((item) => [...item, htmlFiles.includes(item[0])]);

// showing all directory
app.get('/', (req, res) => {
  res.render('directory', {
    siteType: 'home',
    port,
    title: 'All directory',
    folders: allFolders,
  });
});

// get
app.get('/name/:imageFolder', (req, res) => {
  const { imageFolder } = req.params;

  const options = {
    imageFolder,
  };

  if (!existsSync(`./images/${options.imageFolder}`)) {
    res.render('404', {
      siteType: '404',
      port,
      title: locals.title,
      header: locals.header,
      info: locals.info,
      address: locals.address,
    });
    return;
  }

  const allImages = getAllFiles(`./images/${options.imageFolder}/`);

  res.render('template', {
    siteType: 'new',
    port,
    title: imageFolder,
    count: allImages.length,
    images: allImages,
    generate,
    author,
  });
});

// update
app.get('/update/:imageFolder', (req, res) => {
  const { imageFolder } = req.params;
  const options = {
    imageFolder,
  };

  if (!existsSync(`./data/${imageFolder}.json`)) {
    res.render('404', {
      siteType: '404',
      port,
      title: locals.title,
      header: locals.header,
      info: locals.info,
      address: locals.address,
    });
    return;
  }

  const allImages = getAllFiles(`./images/${options.imageFolder}/`);

  const readimg = readJson(`./data/${options.imageFolder}.json`);

  const optionsDate = {
    day: '2-digit',
    month: '2-digit',
    year: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
  };
  let dateCreateJsonFile = statSync(`./data/${imageFolder}.json`).atime;
  dateCreateJsonFile = (new Date(dateCreateJsonFile).toLocaleString('pl', optionsDate));

  res.render('template', {
    siteType: 'update',
    port,
    title: imageFolder,
    data: dateCreateJsonFile,
    count: allImages.length,
    images: readimg,
    generate,
    author,
    // features: readimg,
  });
});

// success
app.get('/success', (req, res) => {
  const htmlText = generate ? 'JSON and HTML' : 'JSON';
  res.render('success', {
    siteType: 'success',
    port,
    title: 'success',
    text: `${htmlText} has been saved`,
  });
});

// 404 redirect
app.use((req, res) => {
  res.status(404)
    .render('404', {
      siteType: '404',
      port,
      title: locals.title,
      header: locals.header,
      info: locals.info,
      address: locals.address,
    });
});

// 500 error
app.use((err, req, res) => {
  res.type('text/plain')
    .status(500)
    .send('500 Server Error');
});

// listen http://localhost:3000
app.listen(port, () => {
  // eslint-disable-next-line no-console
  console.log(`Server is runing at port: http://localhost:${port}`);
});

sources/helper/images.js
const {
  statSync, existsSync, readdirSync, readFileSync, mkdirSync,
} = require('fs');

const getAllFiles = (dir) => readdirSync(dir).reduce((files, file) => {
  const name = `${dir}${file}`;
  const isDirectory = statSync(name).isDirectory();
  return isDirectory ? [...files, ...getAllFiles(name)] : [...files, name];
}, []);

const getAllDirectory = (dir) => readdirSync(dir, { withFileTypes: true })
  .filter((dirent) => dirent.isDirectory())
  .map((dirent) => dirent.name);

const getAllJson = (dir, extension) => readdirSync(dir)
  .filter((file) => file.match(new RegExp(`.*\\.(${extension})`, 'ig')))
  .map((map) => map.replace(`.${extension}`, ''));

const readJson = (dir) => {
  const fileContents = readFileSync(dir, 'utf8');
  const images = JSON.parse(fileContents);
  return images.body.items.map((image) => `${image.path}${image.img}`);
};

const checkFolders = (dir) => {
  try {
    if (!existsSync(dir)) {
      mkdirSync(dir);
    }
  } catch (err) {
    console.error(err);
  }
};


module.exports = {
  checkFolders, getAllFiles, getAllDirectory, getAllJson, readJson,
};


sources/helper/template.js
const { writeFile } = require('fs');

const dateModified = new Date().toISOString().slice(0, 10);

const template = (options) => {
  const {
    nameFolder,
    seoTitle,
    seoDescription,
    bodyTitle,
    bodyDate,
    bodyText,
    bodyAuthor,
    images,
  } = options;

  const date = bodyDate.replace(/"/g, '').split('.');
  const datePublished = `${date[2]}-${date[1]}-${date[0]}`;

  const templateJson = `{
  "head": {
    "title": ${seoTitle},
    "description": ${seoDescription}
  },
  "body": {
    "title": ${bodyTitle},
    "date": ${bodyDate},
    "text": ${bodyText},
    "items": [${images}
    ]
  },
  "schema": {
    "datePublished": "${datePublished}",
    "dateModified": "${dateModified}",
    "author": ${bodyAuthor}
  }
}
  `;

  writeFile(`./data/${nameFolder}.json`, templateJson, (err) => {
    if (err) return console.log(err);
    return console.log('json has been saved');
  });
};

module.exports = template;


app.js
const { existsSync, statSync } = require('fs');
const path = require('path');
const express = require('express');
const compression = require('compression');
const bodyParser = require('body-parser');
const favicon = require('serve-favicon');
const { port, htmlGenerator, author } = require('./config');
// function
const {
  getAllFiles,
  getAllDirectory,
  getAllJson,
  readJson,
  checkFolders,
} = require('./sources/helper/images');

const generate = htmlGenerator === 'true';

const app = express();

// compression static files
app.use(compression());

const fileTemplate = require('./sources/helper/template');
const fileHtml = require('./sources/helper/html');

// time for caching - maxAge
const oneHour = 60 * 1000 * 60;

const locals = {
  title: 'error',
  header: 'UPS !!! error occurred. The reason may be the wrong address.',
  info: 'The address should contain the name of the folder with photos.',
  address: `http://localhost:${port}/<b>name/folder-name</b>`,
};

// load View Engine
app.set('views', path.join(__dirname, 'sources/views'));
app.set('view engine', 'pug');

// form
app.use(bodyParser.urlencoded({ extended: false }));

// path to css, images, favico and html
app.use('/vendor', express.static(path.join(__dirname, 'sources/vendor')));
app.use('/html', express.static(path.join(__dirname, 'html')));
app.use('/images', express.static(path.join(__dirname, 'images'), { maxAge: oneHour }));

// json data
app.use('/update', express.static(path.join(__dirname, 'data')));

// favicon
app.use(favicon(path.join(__dirname, '/public/images/favicon.ico')));

// get post
app.post('/', (req, res) => {
  const {
    folderName,
    seoTitle,
    seoDescription,
    bodyTitle,
    bodyDate,
    bodyText,
    bodyAuthor,
    imageName,
    imageAlt,
    imageText,
    html,
  } = req.body;

  const imagePath = imageName.map((image, index) => `
    {
      "path": "images/${folderName}/",
      "img": "${image}",
      "alt": "${imageAlt[index]}",
      "text": ${JSON.stringify(imageText[index].replace(/\n?\r\n/g, '<br>'))} 
    }`);


  const config = {
    nameFolder: folderName,
    seoTitle: JSON.stringify(seoTitle),
    seoDescription: JSON.stringify(seoDescription),
    bodyTitle: JSON.stringify(bodyTitle),
    bodyDate: JSON.stringify(bodyDate),
    bodyText: JSON.stringify(bodyText),
    bodyAuthor: JSON.stringify(bodyAuthor),
    images: imagePath,
  };

  // save html files, dependent on .env (HTML_GENERATOR=true/false) and checkbox (HTML SAVE)
  if (generate || html === 'on') {
  // save html files
    fileHtml(config);
  }

  // save json filess
  fileTemplate(config);

  res.redirect('./success');
});

// checking if folders exist if they are not created
checkFolders('./data/');
checkFolders('./html/');

const keys = getAllDirectory('./images/');
const jsonFiles = getAllJson('./data/', 'json');
const htmlFiles = getAllJson('./html/', 'html');

const merged = keys.reduce((obj, key) => ({
  ...obj,
  [key]: jsonFiles.includes(key),
}), {});

const allFolders = Object.entries(merged).map((item) => [...item, htmlFiles.includes(item[0])]);

// showing all directory
app.get('/', (req, res) => {
  res.render('directory', {
    siteType: 'home',
    port,
    title: 'All directory',
    folders: allFolders,
  });
});

// get
app.get('/name/:imageFolder', (req, res) => {
  const { imageFolder } = req.params;

  const options = {
    imageFolder,
  };

  if (!existsSync(`./images/${options.imageFolder}`)) {
    res.render('404', {
      siteType: '404',
      port,
      title: locals.title,
      header: locals.header,
      info: locals.info,
      address: locals.address,
    });
    return;
  }

  const allImages = getAllFiles(`./images/${options.imageFolder}/`);

  res.render('template', {
    siteType: 'new',
    port,
    title: imageFolder,
    count: allImages.length,
    images: allImages,
    generate,
    author,
  });
});

// update
app.get('/update/:imageFolder', (req, res) => {
  const { imageFolder } = req.params;
  const options = {
    imageFolder,
  };

  if (!existsSync(`./data/${imageFolder}.json`)) {
    res.render('404', {
      siteType: '404',
      port,
      title: locals.title,
      header: locals.header,
      info: locals.info,
      address: locals.address,
    });
    return;
  }

  const allImages = getAllFiles(`./images/${options.imageFolder}/`);

  const readimg = readJson(`./data/${options.imageFolder}.json`);

  const optionsDate = {
    day: '2-digit',
    month: '2-digit',
    year: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
  };
  let dateCreateJsonFile = statSync(`./data/${imageFolder}.json`).atime;
  dateCreateJsonFile = (new Date(dateCreateJsonFile).toLocaleString('pl', optionsDate));

  res.render('template', {
    siteType: 'update',
    port,
    title: imageFolder,
    data: dateCreateJsonFile,
    count: allImages.length,
    images: readimg,
    generate,
    author,
    // features: readimg,
  });
});

// success
app.get('/success', (req, res) => {
  const htmlText = generate ? 'JSON and HTML' : 'JSON';
  res.render('success', {
    siteType: 'success',
    port,
    title: 'success',
    text: `${htmlText} has been saved`,
  });
});

// 404 redirect
app.use((req, res) => {
  res.status(404)
    .render('404', {
      siteType: '404',
      port,
      title: locals.title,
      header: locals.header,
      info: locals.info,
      address: locals.address,
    });
});

// 500 error
app.use((err, req, res) => {
  res.type('text/plain')
    .status(500)
    .send('500 Server Error');
});

// listen http://localhost:3000
app.listen(port, () => {
  // eslint-disable-next-line no-console
  console.log(`Server is runing at port: http://localhost:${port}`);
});
